mongo/integration/unified/client_entity.go

```diff
@@ -133,21 +176,171 @@ func (c *clientEntity) failedEvents() []*event.CommandFailedEvent {
+			}
+		}
 	}
 }
 
 func (c *clientEntity) processSucceededEvent(_ context.Context, evt *event.CommandSucceededEvent) {
 	if c.getRecordEvents() {
-		c.succeeded = append(c.succeeded, evt)
+		if _, ok := c.observedEvents["commandSucceededEvent"]; ok {
+			c.succeeded = append(c.succeeded, evt)
+		}
+		eventListIDs, ok := c.storedEvents["CommandSucceededEvent"]
+		if ok {
+			bsonBuilder := bsoncore.NewDocumentBuilder().
+				AppendString("name", "CommandSucceededEvent").
+				AppendInt64("observedAt", getCurrentTime()).
+				AppendString("commandName", evt.CommandName).
+				AppendInt64("requestId", evt.RequestID).
+				AppendString("connectionId", evt.ConnectionID)
+			if evt.ServerID != nil {
+				bsonBuilder.AppendString("serverId", evt.ServerID.String())
+			}
+			for _, id := range eventListIDs {
+				c.entityMap.appendEventsEntity(id, bson.Raw(bsonBuilder.Build()))
+			}
+		}
 	}
 }
 
 func (c *clientEntity) processFailedEvent(_ context.Context, evt *event.CommandFailedEvent) {
 	if c.getRecordEvents() {
-		c.failed = append(c.failed, evt)
+		if _, ok := c.observedEvents["commandFailedEvent"]; ok {
+			c.failed = append(c.failed, evt)
+		}
+		eventListIDs, ok := c.storedEvents["CommandFailedEvent"]
+		if ok {
+			bsonBuilder := bsoncore.NewDocumentBuilder().
+				AppendString("name", "CommandFailedEvent").
+				AppendInt64("observedAt", getCurrentTime()).
+				AppendInt64("durationNanos", evt.DurationNanos).
+				AppendString("commandName", evt.CommandName).
+				AppendInt64("requestId", evt.RequestID).
+				AppendString("connectionId", evt.ConnectionID)
+			if evt.ServerID != nil {
+				bsonBuilder.AppendString("serverId", evt.ServerID.String())
+			}
+			bsonBuilder.AppendString("failure", evt.Failure)
+			for _, id := range eventListIDs {
+				c.entityMap.appendEventsEntity(id, bson.Raw(bsonBuilder.Build()))
+			}
+		}
+	}
+}
+
+func getPoolEventDocument(evt *event.PoolEvent) bson.Raw {
+	bsonBuilder := bsoncore.NewDocumentBuilder().
+		AppendString("name", evt.Type).
+		AppendInt64("observedAt", getCurrentTime()).
+		AppendString("address", evt.Address)
+	if evt.ConnectionID != 0 {
+		bsonBuilder.AppendString("connectionId", fmt.Sprint(evt.ConnectionID))
+	}
+	if evt.PoolOptions != nil {
+		optionsDoc := bsoncore.NewDocumentBuilder().
+			AppendString("maxPoolSize", fmt.Sprint(evt.PoolOptions.MaxPoolSize)).
+			AppendString("minPoolSize", fmt.Sprint(evt.PoolOptions.MinPoolSize)).
+			AppendString("maxIdleTimeMS", fmt.Sprint(evt.PoolOptions.WaitQueueTimeoutMS)).
+			Build()
+		bsonBuilder.AppendDocument("poolOptions", optionsDoc)
+	}
+	if evt.Reason != "" {
+		bsonBuilder.AppendString("reason", evt.Reason)
+	}
+	return bson.Raw(bsonBuilder.Build())
+}
+
+func (c *clientEntity) processPoolEvent(evt *event.PoolEvent) {
+	eventBSON := getPoolEventDocument(evt)
+	switch evt.Type {
+	// TODO GODRIVER-1827: add storePoolReady and storeConnectionCheckOutStarted events
+	case event.PoolCreated:
```

Comment:
nit: The loop to store the events is duplicated in each case statement. Would it be clearer to do:
```
eventBSON := ...
var eventType string

switch evt.Type {
case event.PoolCreated:
    eventType = "PoolCreatedEvent"
...
}

eventListIDs, ok := c.storedEvents[eventType]
...
```